lib/dd.ml

```diff
@@ -0,0 +1,2719 @@
+       ; 0x0D; 0x8D; 0x4D; 0xCD; 0x2D; 0xAD; 0x6D; 0xED
+       ; 0x1D; 0x9D; 0x5D; 0xDD; 0x3D; 0xBD; 0x7D; 0xFD
+       ; 0x03; 0x83; 0x43; 0xC3; 0x23; 0xA3; 0x63; 0xE3
+       ; 0x13; 0x93; 0x53; 0xD3; 0x33; 0xB3; 0x73; 0xF3
+       ; 0x0B; 0x8B; 0x4B; 0xCB; 0x2B; 0xAB; 0x6B; 0xEB
+       ; 0x1B; 0x9B; 0x5B; 0xDB; 0x3B; 0xBB; 0x7B; 0xFB
+       ; 0x07; 0x87; 0x47; 0xC7; 0x27; 0xA7; 0x67; 0xE7
+       ; 0x17; 0x97; 0x57; 0xD7; 0x37; 0xB7; 0x77; 0xF7
+       ; 0x0F; 0x8F; 0x4F; 0xCF; 0x2F; 0xAF; 0x6F; 0xEF
+       ; 0x1F; 0x9F; 0x5F; 0xDF; 0x3F; 0xBF; 0x7F; 0xFF |]
+    in t.(bits)
+  [@@inline]
+
+  let fixed_lit, fixed_dist =
+    let tbl_lit =
+      Array.init 288 @@ fun n ->
+      if n < 144 then 8
+      else if n < 256 then 9
+      else if n < 280 then 7
+      else 8 in
+    let tbl_dist =
+      let res = Array.make (1 lsl 5) 0 in
+      Array.iteri (fun i _ -> res.(i) <- (5 lsl 15) lor reverse_bits (i lsl 3)) res ; res in
+    let tbl_lit, max_lit = huffman LENS tbl_lit 0 288 in
+    Lookup.make tbl_lit max_lit, Lookup.make tbl_dist 5
+
+  let checksum d =
+    Window.checksum d.w
+
+  let rec flat d =
+    let len = min (min (i_rem d) d.l) (bigstring_length d.o - d.o_pos) in
+    Window.blit d.w d.i d.i_pos d.o d.o_pos len ;
+
+    d.o_pos <- d.o_pos + len ;
+    d.i_pos <- d.i_pos + len ;
+    d.l <- d.l - len ;
+
+    if d.l == 0
+    then ( if d.last
+           then ( d.s <- End_of_inflate ; K )
+           else ( d.s <- Header ; K ) )
+    else match i_rem d, bigstring_length d.o - d.o_pos with
+      | 0, _ ->
+        ( match d.src with
+          | `String _ -> eoi d ; err_unexpected_end_of_input d
+          | `Channel ic ->
+            let len = input_bigstring ic d.i 0 (bigstring_length d.i) in
+            src d d.i 0 len ; flat d (* XXX(dinosaure): check this branch. TODO! *)
+          | `Manual -> Await)
+      | _, 0 -> Flush
+      | _, _ -> assert false
+
+  let flat_header d =
+    let k d =
+      let t_pos = ref 0 in
+      let hold = ref d.hold in
+      let bits = ref d.bits in
+      let len = ref 0 and nlen = ref 0xffff in
+
+      let consume () =
+        if !bits < 8
+        then ( hold := ((unsafe_get_uint8 d.t !t_pos) lsl !bits) lor !hold
+             ; bits := !bits + 8
+             ; incr t_pos ) in
+
+      consume () ;
+      len := !hold land 0xff ;
+      hold := !hold asr 8 ;
+      bits := !bits - 8 ;
+
+      consume () ;
+      len := ((!hold land 0xff) lsl 8) lor !len ;
+      hold := !hold asr 8 ;
+      bits := !bits - 8 ;
+
+      consume () ;
+      nlen := !hold land 0xff ;
+      hold := !hold asr 8 ;
+      bits := !bits - 8;
+
+      consume () ;
+      nlen := ((!hold land 0xff) lsl 8) lor !nlen ;
+      hold := !hold asr 8 ;
+      bits := !bits - 8 ;
+
+      if !nlen != 0xffff - !len
+      then err_invalid_complement_of_length d
+      else ( d.hold <- 0 ; d.bits <- 0 ; d.l <- !len ; d.s <- Flat ; flat d ) in
+    let bits = d.bits in
```

Comment:
This is kind of hard to read (`d.bits` vs `bits`).
Do we need to handle negative values for `d.bits`, ie can we do something like this and shave off 16 instructions (GodBolt: [this](https://ocaml.godbolt.org/z/MllVxy) vs [current](https://ocaml.godbolt.org/z/S-ecYZ))?
```ocaml
    d.hold <- d.hold asr (d.bits land 7) ;
    (* diff between d.bits and d.bits rounded down to nearest multiple of 8 *)

    let truncated_bits = d.bits land (lnot 7) in
    (* round down to nearest multiple of 8, logical equivalents:
       d.bits land (lnot (8-1))
       d.bits land (lnot 7)
     *)
    (* for some reason saving this variable locally instead of accessing d.bits twice
       shaves off one instruction when compiling with flambda *)

    d.bits <- truncated_bits;

    let required = 4 - (truncated_bits asr 3) in
    d.s <- Flat_header ;
    t_need d required ; t_fill k d
```

Reply:
Hmmhmm, I like your implementation but on my side, it's difficult to understand (specially about `lnot 7`. So I will add your implementations with comments :+1: .

Reply:
Your code passes tests 🎉

Reply:
We want to round **down** to the nearest multiple of 8. Eight is itself a power of 2, so we can achieve this with cheap bitmasking instead of expensive `idivq` instructions. That only works when our inputs are positive integers since the negative integers look a bit different.

 `lnot 7` flips all the bits in the number `7` (up to word size), so if you have a byte:
```
0b00 000 111 = 7
The logical NOT will be:
0b11 111 000 = (2^8 - 1) - 7 = 248 (unsigned) = (-8) (signed)

example of b = a land (lnot 7):
= 0b00 11 0 101 = a = 53
& 0b11 11 1 000
= 0b00 11 0 000 = b = 48
```

`lnot 7` will be computed by the compiler, so `let truncated_bits = d.bits land (lnot 7)` will just turn into an AND where all bits are `1` except the three rightmost (the `0b111` from `7`) which are `0`. So if we AND with that, we mask out the numbers that are not a multiple of 8, and thus round down. :)